src/crypto/rfc8439.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <crypto/rfc8439.h>
+
+#include <crypto/chacha20.h>
+#include <crypto/common.h>
+
+#include <cstring>
+
+#ifndef RFC8439_TIMINGSAFE_BCMP
+#define RFC8439_TIMINGSAFE_BCMP
+
+int rfc8439_timingsafe_bcmp(const unsigned char* b1, const unsigned char* b2, size_t n)
+{
+    const unsigned char *p1 = b1, *p2 = b2;
+    int ret = 0;
+
+    for (; n > 0; n--)
+        ret |= *p1++ ^ *p2++;
+    return (ret != 0);
+}
+
+#endif // RFC8439_TIMINGSAFE_BCMP
+
+inline size_t padded16_size(size_t len)
+{
+    return (len % 16 == 0) ? len : (len / 16 + 1) * 16;
+}
+
+void ComputeRFC8439Tag(const std::array<std::byte, POLY1305_KEYLEN>& polykey,
+                       Span<const std::byte> aad, Span<const std::byte> ciphertext,
```

Comment:
nit: const-correctness
```suggestion
                       const Span<const std::byte> aad, const Span<const std::byte> ciphertext,
```